### PR TITLE
Closes #24

### LINKED ISSUE #24: Extract data from production ntuples on the Grid
Extend the fill_spectrum code so it is compatable with production ntuples located on the grid not just root files generated by rat.

### PR DESCRIPTION
**Pull request by arushanova**
Added two functions to `fill_spectrum` to get the mc and reco data from ntuples

### CODE REVIEW

**Comment by jwaterfield** [comment]
Seems to work by eye but can't look at output spectrums as there seems to be bugs in the plotting code @pgjones:

Traceback (most recent call last):
  File "test_evelina.py", line 14, in <module>
    plot.plot_projection(loaded_mc,0)
  File "/home/james/echidna/echidna/output/plot.py", line 30, in plot_projection
    x = _produce_axis(spectra._energy_low, spectra._energy_high, spectra._energy_bins)
AttributeError: 'str' object has no attribute '_energy_low'

**Comment by jwaterfield** [comment]
Could do with pep8 modifications theres quite a few lines which fail pep8 checks

File: echidna/core/fill_spectrum.py [hunk 1]
```diff
@@ -132,3 +133,95 @@ def fill_mc_spectrum(filename, spectrumname, T):
...
+    print spectrumname
+    chain=TChain("output")
+    chain.Add(filename)
+    spectrum = spectra.Spectra(str(spectrumname), chain.GetEntries())
+
+    times = [0]
+    for time_step in range(0, spectrum._time_bins):
+        time = time_step * spectrum._time_width + spectrum._time_low 
+        times.append(time)
+
```
Comment (jwaterfield) [inline echidna/core/fill_spectrum.py]: Hmmm maybe I am missing something but I thought a random time should be assigned to each event? This appears to be just linearly assigning times from start -> end

**Comment by jwaterfield** [comment]
Ignore my earlier comments @pgjones the :bug: was in my code :dizzy_face:

**Comment by arushanova** [comment]
Things that could be:

1) fill_reco_spectrum can become fill_mc_spectrum or fill_reco_spectrum or fill_mc_ntuple_spectrum or fill_reco_ntuple_spectrum

2) "Dr Majoron AV" - the name of the future spectrum. If there is AV in the name, than it will be applied the timing for AV backgrounds, if there is no AV in the name, than the timing is for normal backgrounds.


This code should be in the major echidna directory. At least it work for me this way.


import echidna.core.spectra as spectra
import echidna.output.plot_root as plot
import echidna.core.fill_spectrum as fill_spectrum

filename = "/data/snoplus/evelina/PileUpGeneration/RAT5/root_files/Te130_0n2b_n1.root"

t = fill_spectrum.fill_reco_spectrum( filename, "Dr Majoron AV", 100.0 )
print t
plotter = plot.plot_projection( t, 2)

raw_input("RET to EXIT")




Cheers,

Evelina





________________________________

File: echidna/core/fill_spectrum.py (hunk 1 above)
Comment (jwaterfield) [inline echidna/core/fill_spectrum.py]: Also appears to be an extra 0 in your list

**Comment by jwaterfield** [comment]
Yeah sorry I had a bug in my plotting code D'oh. The reco ntuple appears to be saving the spectrum fine but don't seem to be able to get anything in the mc ntuple. Haven't worked out why yet... Have you tried this?

**Comment by ashleyrback** [comment]
Does `event.scintFit` in the ntuple tell you if both energy and position reconstruction has passed?

**Comment by jwaterfield** [comment]
Yep this should be added

**Comment by arushanova** [comment]
I guess... Isn't the scintFit take care of both energy and position? Do you think its wrong...?  In root file there was the possibility to check the default fitter (because in future there might be tellurium fitter), with ntuple I haven't found this.

**Comment by arushanova** [comment]
It seems Ashley can't run the code either. But it is fine for me.... no error messages, no warning.

**Comment by ashleyrback** [comment]
It's working fine for me now. I get plots for both reco and mc.

File: echidna/core/fill_spectrum.py (hunk 1 above)
Comment (ashleyrback) [inline echidna/core/fill_spectrum.py]: @jwaterfield the range of times is correct (without the extra zero - now fixed). Each event is copied into all time bins but with different weights applied.

**Comment by jwaterfield** [comment]
I still cant get the mc working... No error messages just blank plots. Could you send me your ntuple and script @ashleyrback?

Yeah I guess variables in ntuples will change as well as time goes by. We will just have to cross that bridge when it comes to it.

**Comment by jwaterfield** [comment]
Hmm works with your ntuple but not my one which I got from the grid? There must be some reason why this ntuple isn't working. Have you tried it with ntuples from the grid? I'll email you one I have.

**Comment by arushanova** [comment]
There is one line in my code, that checks if fits is ok for reco root file. Ashley and pep8 both think that these "not" conditions should be on separate lines. I disagree, I think this check is quite basic and one doesn't need to stare at it, just look through it. When it is spread for 3 lines it draws the attention. So I like the long line and want to keep it. If people keen on changing it - I will.

**Comment by arushanova** [comment]
I'll check the code with ntuple from grid. I think we do have them in QMUL.

**Comment by arushanova** [comment]
James, there are no ntuples for rat 5 on the GRID. I've simulated ntuple with rat 5 myself just to test. Which ntuples are you using?

**Comment by jwaterfield** [comment]
As far as I'm aware there are no 5.0 ntuples. I was testing with 4.5

**Comment by ashleyrback** [comment]
@jwaterfield I've just tested on the RAT-4.5.0 Bi210 production ntuple and it worked fine. I think the ntuple structure didn't change during the DS review, so @arushanova's ntuple functions should work fine for 4.5 data. They won't work for 4.5 root files, but I don't think any of us need to use them at the moment. I'll send you  the script I just used James.

**Comment by jwaterfield** [comment]
Strange I was using the script Evelina sent and it was plotting reco fine just not mc.  Also the ntuple you sent me worked fine. I will send you my ntuple.

**Comment by jwaterfield** [comment]
Yep the script you sent doesnt work on that ntuple I've just sent you. Are you sure you tested the mc as well as the reco. Reco is fine (in your script) its just mc that has an  issue.

**Comment by jwaterfield** [comment]
As I said in my email the reason the mc plot was failing was because the default values in the histogram only include positions with radius within the AV. I propose we extend the default values to include all posible positions within the detector as default. People can then cut based on radius as they wish afterwards or change the default values before they begin filling. Could you change the line in the spectra class @arushanova and I will go ahead and merge.

**Comment by jwaterfield** [comment]
The cavity has a r_max of 10000 mm. I guess this will be sufficient unless we need to extend to the rock?

**Pull request merged**